toolchain/semantics/semantics_parse_tree_handler.cpp

```diff
@@ -451,23 +448,75 @@ auto SemanticsParseTreeHandler::HandlePackageLibrary(
   CARBON_FATAL() << "TODO";
 }
 
+auto SemanticsParseTreeHandler::SaveParam() -> bool {
+  // Copy the last node added to the IR block into the params block.
+  auto ir_id = node_block_stack_.Peek();
+  if (!ir_id.is_valid()) {
+    return false;
+  }
+  auto& ir = semantics_->GetNodeBlock(ir_id);
+  CARBON_CHECK(!ir.empty())
+      << "Should only have a valid ID if a node was added";
+  auto& param = ir.back();
+  auto& params = semantics_->GetNodeBlock(params_stack_.PeekForAdd());
+  if (!params.empty() && param == params.back()) {
+    // The param was already added after a comma.
+    return false;
+  }
+  params.push_back(ir.back());
+  return true;
+}
+
 auto SemanticsParseTreeHandler::HandleParameterList(ParseTree::Node parse_node)
     -> void {
-  // TODO: This should transform into a usable parameter list. For now
-  // it's unused and only stored so that node counts match.
-  node_stack_.PopAndDiscardSoloParseNode(ParseNodeKind::ParameterListStart);
-  node_stack_.Push(parse_node);
+  // If there's a node in the IR block that has yet to be added to the params
+  // block, add it now.
+  SaveParam();
+
+  while (true) {
+    switch (auto parse_kind =
+                parse_tree_->node_kind(node_stack_.PeekParseNode())) {
+      case ParseNodeKind::ParameterListStart:
+        node_stack_.PopAndDiscardSoloParseNode(
+            ParseNodeKind::ParameterListStart);
+        finished_params_stack_.push_back(
+            {node_block_stack_.Pop(), params_stack_.Pop()});
+        node_stack_.Push(parse_node);
+        return;
+
+      case ParseNodeKind::ParameterListComma:
+        node_stack_.PopAndDiscardSoloParseNode(
+            ParseNodeKind::ParameterListComma);
+        break;
+
+      case ParseNodeKind::PatternBinding:
+        node_stack_.PopAndDiscardId(ParseNodeKind::PatternBinding);
+        break;
+
+      default:
+        // This should only occur for invalid parse trees.
+        CARBON_FATAL() << "TODO: " << parse_kind;
+    }
+  }
+
+  llvm_unreachable("loop always exits");
 }
 
 auto SemanticsParseTreeHandler::HandleParameterListComma(
-    ParseTree::Node /*parse_node*/) -> void {
-  CARBON_FATAL() << "TODO";
+    ParseTree::Node parse_node) -> void {
+  node_stack_.Push(parse_node);
+
+  // Copy the last node added to the IR block into the params block.
+  CARBON_CHECK(SaveParam())
```

Comment:
I think putting state changes in a `CARBON_CHECK` harms readability -- when I'm scanning the code, I will generally skip over `CARBON_CHECK`s because I assume they're just checking invariants. Can you move the `SaveParam` call out of the check?
```suggestion
  bool had_param_before_comma = SaveParam();
  CARBON_CHECK(had_param_before_comma)
```